business/mesh.go

```diff
@@ -54,47 +57,49 @@ func NewMeshService(
 }
 
 func (in *MeshService) CanaryUpgradeStatus() (*models.CanaryUpgradeStatus, error) {
-	upgrade := in.conf.ExternalServices.Istio.IstioCanaryRevision.Upgrade
-	current := in.conf.ExternalServices.Istio.IstioCanaryRevision.Current
-	migratedNsList := []string{}
-	pendingNsList := []string{}
-
-	// If there is no canary configured, return empty lists
-	if upgrade == "" {
-		return &models.CanaryUpgradeStatus{MigratedNamespaces: migratedNsList, PendingNamespaces: pendingNsList}, nil
-	}
-
-	// Get migrated and pending namespaces
-	// TODO: Support multi-primary
-	migratedNss, err := in.homeClusterSAClient.GetNamespaces(fmt.Sprintf("istio.io/rev=%s", upgrade))
+	kubeCache, err := in.kialiCache.GetKubeCache(in.conf.KubernetesConfig.ClusterName)
 	if err != nil {
 		return nil, err
 	}
-	for _, ns := range migratedNss {
-		migratedNsList = append(migratedNsList, ns.Name)
-	}
 
-	pendingNss, err := in.homeClusterSAClient.GetNamespaces(fmt.Sprintf("%s=enabled", in.conf.IstioLabels.InjectionLabelName))
+	revisions, err := istio.GetHealthyIstiodRevisions(kubeCache, in.conf.IstioNamespace)
 	if err != nil {
 		return nil, err
 	}
-	for _, ns := range pendingNss {
-		pendingNsList = append(pendingNsList, ns.Name)
-	}
+	namespacesPerRevision := make(map[string][]string)
 
-	pendingNss, err = in.homeClusterSAClient.GetNamespaces(fmt.Sprintf("%s=%s", in.conf.IstioLabels.InjectionLabelRev, current))
-	if err != nil {
-		return nil, err
-	}
-	for _, ns := range pendingNss {
-		pendingNsList = append(pendingNsList, ns.Name)
+	if len(revisions) == 1 {
+		return &models.CanaryUpgradeStatus{
+			NamespacesPerRevision: namespacesPerRevision,
+		}, nil
 	}
 
+	for _, revision := range revisions {
+		nsList := make(map[string]bool)
+		// Get namespaces for revision
+		// TODO: Support multi-primary
+		nss, err := in.homeClusterSAClient.GetNamespaces(fmt.Sprintf("%s=%s", in.conf.IstioLabels.InjectionLabelRev, revision))
```

Comment:
This probably has perf implications and I don't think it will work when `cluster_wide_access = false` but we can update this once we add support for multi-primary.